### ISSUE TITLE
Scripts and Styles in components included inside mdx content, loaded through another component, don't work

### ISSUE THREAD

**Issue by trafnar** [withastro/astro#7761]
### What version of `astro` are you using?

2.9.1

### Are you using an SSR adapter? If so, which one?

No

### What package manager are you using?

npm

### What operating system are you using?

Mac

### What browser are you using?

Chrome

### Describe the Bug

I'm looping through a content collection, using a component to render each content item. The content items themselves are defined in .mdx files, each with a component inside the mdx. That component does not get styling/scripts applied.

[Reproducible example Stackblitz project](https://stackblitz.com/edit/github-u2yta7?file=src%2Fpages%2Findex.astro)

<details><summary>Basic Example of what's in the linked project</summary>
<p>

index.astro (some imports removed for brevity)
```astro
---
const posts = (await getCollection("posts"));
---
{posts.map((post) => <Post post={post} />)}
```

components/Post.astro
```astro
---
const { post } = Astro.props;
const { Content } = await post.render();
---

<h2>
  {post.data.title}
</h2>

<Content />
```

components/Card.astro
```astro
<div class="card">This is a card</div>
<style>
.card{background-color:yellow; border:1px solid red;}
</style>
```

content/posts/my-post.mdx
```mdx
---
title: "My Post"
---
import Card from '../../components/Card.astro';
<Card />
```

The `<Card />` used in `my-post.mdx` will not have styling applied

</p>
</details> 

### What's the expected result?

The `<Card />` component should have it's local styles applied.

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/github-u2yta7?file=src%2Fpages%2Findex.astro

### Participation

- [X] I am willing to submit a pull request for this issue.

**Comment by lilnasy** [comment]
Thank you for the clean reproduction, it was really helpful!

I'll let one of the other maintainers confirm, but this seems to be an architectural limitation.

Due to Astro's streaming, we only wait until the page's frontmatter has run before generating the `<head>` element. Frontmatter of a component (in your case, `components/Post.astro`) cannot add scripts or stylesheets because by the time it's run, the `<head>` is already sent to the browser so that it can start loading assets.

In case of SSG, it shouldn't matter, but the same architecture is used for both modes.

**Comment by ChrisPlease** [comment]
Just started exploring Astro yesterday and ran into this today. My component is a 1 off, so as a work around I set `<style is:inline>`, but I can see how that's not useful in a loop.

**Comment by kurtextrem** [comment]
If anyone finds this issue: I've worked around it by including the component in an Astro layout (`.astro`), so the price we pay is duplicate HTML output (1x in the MDX, 1x in the layout), but assets (CSS & JS) are only added once.

I still think this is a kind of drastic limitation and should definitely be documented somewhere.

**Comment by natemoo-re** [comment]
Duplicate issues
- https://github.com/withastro/astro/issues/7426
- https://github.com/withastro/astro/issues/6434

Closing those to keep the conversation going here.

**Comment by natemoo-re** [comment]
Thanks for investigating @lilnasy! I'm not convinced this is an architectural limitation—it might just be a bug with how we crawl the module graph for styles. 

There's currently some effort to improve that, but we had to move it behind an experimental flag for the time being. See some of the discussion in https://github.com/withastro/astro/issues/7857

**Comment by lilnasy** [comment]
You can add codesandbox or a github repo in the reproduction.

**Comment by trafnar** [comment]
@lilnasy good idea. I believe this issue is actually distinct so I've opened an issue with a reproducible example as a github repo #8436

**Comment by haikyuu** [comment]
@trafnar In development, styles are loaded using an import like so `import "/src/imba-component.imba?imba&type=style&lang.css";`. It's because we're defaulting to `emitCss = true`.
So it's working for both content collections and pages. (I think #8436 can be closed)

But in the build, styles are NOT included in content collection components. [Stackblitz link to reproduce ](https://stackblitz.com/~/github.com/haikyuu/astro-imba-issue)


<img width="1727" alt="image" src="https://github.com/withastro/astro/assets/8558836/75aa64ff-bbaa-4a55-85c0-cd39479892ef">

whereas they are included in regular pages
<img width="1726" alt="image" src="https://github.com/withastro/astro/assets/8558836/f0a87b3a-c951-4405-b3ce-0b20476c411b">


So, from here it seems like Astro forgets to add the link tag in content collections 🤔

**Comment by matthewp** [comment]
This might be the same cause as https://github.com/withastro/astro/issues/8399. I'm working on that one now.

**Comment by trafnar** [comment]
@matthewp can't tell you how much I appreciate you working on this issue ❤️

**Comment by lilnasy** [comment]
Wrote up a proposal based on `@matthewp`'s idea: https://github.com/withastro/roadmap/discussions/707.

**Comment by lilnasy** [comment]
I have published an integration that fixes this issue. [GitHub](https://github.com/lilnasy/gratelets/tree/main/packages/adds-to-head) | [NPM](https://www.npmjs.com/package/astro-adds-to-head)

I have tested it to be working against the original repro provided by **trafnar** (https://stackblitz.com/edit/github-u2yta7?file=src%2Fpages%2Findex.astro)

**Comment by matthewp** [comment]
I think this is more of a limitation of how Content Collections works and probably not something that is fixable. However this feature might make this possible: https://caniuse.com/mdn-html_elements_link_blocking

**Comment by Trombach** [comment]
Hi @lilnasy, I had been using your integration (not directly but I copied the code file, since svelte-check had an issue with the integration for some reason 🤔 ), thanks for that btw! I believe the workaround is not working anymore since the MDX 3.0 integration was released :(

**Comment by jcayzac** [comment]
I just came to say I'm using SSG and I do see this issue.

**Comment by jcayzac** [comment]
As a workaround, I include all the components available to .mdx pages in my layout, wrapped by a `<template>` element. This has Astro include the styles in `<head>` when building.

**Comment by kyleshevlin** [comment]
I have found a workaround that works for me, but I know won't work for everyone.

In the file where you call `entry.render()` to get the `Content` component, pass in the components that are giving you trouble in to the `components` prop on `Content`. I'll use my blog's code for example:

```astro
---
import TroublesomeComponent from './components/TroublesomeComponent.astro'

// other stuff

const { Content: PostContent } = await entry.render()
---

<PostContent components={{ TroublesomeComponent }} />
```

This now makes `TroublesomeComponent` globally available to any `PostContent` I write, which then adds the script and styles to the `<head>`.

Again, this might not be ideal, but I can live with this workaround and thought it might help others.

**Comment by snapwich** [comment]
Hey Astro team, I really like what you've made here and just wanted to chime in here on how important I think it is to support this feature. 

Currently [I have a blog](https://www.richsnapp.com) that is written in Nuxt 2. The reason I chose Nuxt was because I wanted to write blog posts that could include technical demos. Since Nuxt runs on Vue I can create can create "islands" that run various things as part of my posts like I did [here](https://www.richsnapp.com/blog/2020/9-25-rebalancing-your-lazy-portfolio). At the time, Nuxt didn't come with any easy way to automatically aggregate my blog posts into an index page or generate my dynamic tag pages so I ended up writing my own webpack loader that does something similar to your content collections.

Now with Nuxt moving away from webpack and also with me wanting to write technical demos in more than just Vue I figured I could either work on refactoring my webpack loader into vite and upgrade everything to Nuxt 3 + Vue 3 while figuring out how to embed React/Svelte/Solid islands into Nuxt, or I could investigate moving to a new solution entirely since I had a lot of work ahead of me anyways. Enter Astro.

I'd heard about Astro before and knew that it supported a component "island" concept with multiple frameworks so this was the first place I looked. Then after digging into the documentation I saw that it supported MDX content collections that behaved very similar to the way I was using [my vue single-file component content pages](https://github.com/snapwich/richsnapp.com/blob/master/blog/2018/01-07-we-have-a-blog/index.vue) and having them aggreggated. Basically my webpack loader parses the exports and pulls out the "frontmatter", the template content gets rendered as markdown and since it's a vue single-file component i can add styles and components inside as well.

Needless to say, Astro seemed to check all the boxes I needed so I got to work this weekend refactoring my blog to Astro. I'm on day two now and got everything working _until_ I hit my first blog post that had custom styling and components which is how I discovered this thread.

Given that component islands and content collections are such highly-promoted features of Astro I am honestly very surprised to find out that they don't work together... Even re-reading the docs it still doesn't seem clear to me that that is the case, really there is only [this small blurb](https://docs.astro.build/en/guides/styling/#markdown-styling) that seems to suggest you may need to find alternate solutions when using markdown. But this github issue seems to confirm that they indeed do not work together.

**TLDR;** So anyways, sorry for the wall of context, but basically what I want to say with all this is that if this is indeed an architectural limitation of Astro I feel it should be highlighted prominently in both the Content Collection and Components/Islands sections of your documentation. I don't think I was making a big logical leap as I was reading the docs to assume that they would work together and I am also assuming most new users of Astro will make the same assumption unless the documentation strongly tells them that this is not the case.

Also, after reading this thread, I couldn't quite tell if making these features work together was going to be on your future roadmap. I read [the linked proposal](https://github.com/withastro/roadmap/discussions/707) and couldn't quite understand if that would fix this issue. If it's not on your roadmap I would strongly suggest making components with content collections as a high-priority item; not just because I want it, but because I assume most of your users who were drawn to Astro for the same reasons I was will want it and not supporting it seems like a huge oversight. I realize things like this happen as a project grows and matures (I think content collections came much later from reading your docs?) and I also realize this is an open source project, so please understand that I am not trying to chastise but instead stressing how important I think supporting this is for the success of Astro even if it means re-architecting some things! For now I'll be putting my blog refactor on the backburner as this seems like a blocker that, for me, doesn't have a workaround (each of my pages can include a lot of code and styles that are really independent of the other pages, so exposing things globally is really a non-solution).

Thanks again for your hard work you've put into this project, let me know if I can help in some way.

**Comment by ascorbic** [comment]
This seems to be fixed when you use content layer: https://stackblitz.com/edit/github-u2yta7-zmksav?file=src%2Fcontent%2Fconfig.ts

I'm going to close this, but if it can still be reproduced with astro@beta then please reply with a reproduction.

![image](https://github.com/user-attachments/assets/dcc6a2c2-ed80-47c5-bdf8-85d22651b980)

**Comment by kurtextrem** [comment]
Seems fixed to me too, *but* you have to ensure you use the following syntax:
```js
import { render } from "astro:content";
const { Content } = await render(post);
```

While using `const { Content } = await post.render();` seems to keep the issue around: https://stackblitz.com/edit/github-u2yta7-shnm1x?file=src%2Flayouts%2FPost.astro (make sure to run `npm run build && npm run preview`).
 Unsure if expected or not, so meanwhile not opening a new issue (since I don't know if `post.render` is deprecated or not).

**Comment by ascorbic** [comment]
Yes, it's deprecated for content layer

**Comment by kurtextrem** [comment]
Awesome, thank you for confirming and fixing this!

**Comment by snapwich** [comment]
just chiming in to say that the latest version of astro resolved the issue i was seeing above and i was able to finish my refactor of [my blog](richsnapp.com) from nuxt to astro. thanks so much for the great work you are all doing here!

**Comment by kitschpatrol** [comment]
Just ran into this in Astro 6. I found that calling `render()` in frontmatter works correctly, but calling `render()` in the template will resurface this bug, where CSS imported in the component are missing. I doubt there's an expectation that the template approach below should work, but it was working fine in dev so I didn't think twice...

---

❌ Template approach
```astro
---
import { render } from 'astro:content'
import { getCollection } from 'astro:content'

const posts = await getCollection('blog')
---

{posts.map(async (post) => {
  const { Content } = await render(post)
  return (
    <article>
      <h2>{post.data.title}</h2>
      <Content />
    </article>
  )
})}
```

---

✅ Frontmatter approach
```astro
---
import { render } from 'astro:content'
import { getCollection } from 'astro:content'

const posts = await getCollection('blog')
const rendered = await Promise.all(
  posts.map(async (post) => {
    const { Content } = await render(post)
    return { post, Content }
  })
)
---

{rendered.map(({ post, Content }) => (
  <article>
    <h2>{post.data.title}</h2>
    <Content />
  </article>
))}
```

---

**Comment by jcayzac** [comment]
@kitschpatrol your non-working version evaluates to an array of promises. I'm surprised any markup is rendered at all. I would expect the rendered HTML to be a series of `[object Promise]`.

**Comment by kitschpatrol** [comment]
@jcayzac Thanks for taking a look, and maybe I'm mistaken, but I believe it's documented that Astro's template rendering logic [queues and resolves promises to allow for streaming](https://docs.astro.build/en/recipes/streaming-improve-page-performance/#including-promises-directly).
